daemon/src/main/java/bisq/daemon/grpc/GrpcOffersService.java

```diff
@@ -86,21 +90,53 @@ public void getOffers(GetOffersRequest req,
     @Override
     public void createOffer(CreateOfferRequest req,
                             StreamObserver<CreateOfferReply> responseObserver) {
-        TransactionResultHandler resultHandler = transaction -> {
-            CreateOfferReply reply = CreateOfferReply.newBuilder().setResult(true).build();
+        CountDownLatch latch = new CountDownLatch(1);
+        try {
+            TransactionResultHandler resultHandler = transaction -> {
+                latch.countDown();
```

Comment:
Why use a CountDownLatch here instead continuing the code in the result handler? If there is n o strong reason for that can we stick to the style like it used elsewhere in the app?

Reply:
I need to wrap the return value in a `OfferInfo` proto before returning it to the CLI.  That's why I used a latch.  Is there a way to get at the offer return value in the TransactionResultHandler?

Reply:
In `GrpcOffersService`, there is a comment explaining why `OfferInfo` is used:

        // The client cannot see bisq.core.Offer or its fromProto method.
        // We use the lighter weight OfferInfo proto wrapper instead, containing just
        // enough fields to view and create offers.

Reply:
I meant to move all to the handler directly.
```
TransactionResultHandler resultHandler = transaction -> {
Offer offer = coreApi.createOffer(
                    req.getCurrencyCode(),
                    req.getDirection(),
                    req.getPrice(),
                    req.getUseMarketBasedPrice(),
                    req.getMarketPriceMargin(),
                    req.getAmount(),
                    req.getMinAmount(),
                    req.getBuyerSecurityDeposit(),
                    req.getPaymentAccountId(),
                    resultHandler);

            OfferInfo offerInfo = new OfferInfo.OfferInfoBuilder()
                    .withId(offer.getId())
                    .withDirection(offer.getDirection().name())
                    .withPrice(offer.getPrice().getValue())
                    .withUseMarketBasedPrice(offer.isUseMarketBasedPrice())
                    .withMarketPriceMargin(offer.getMarketPriceMargin())
                    .withAmount(offer.getAmount().value)
                    .withMinAmount(offer.getMinAmount().value)
                    .withVolume(offer.getVolume().getValue())
                    .withMinVolume(offer.getMinVolume().getValue())
                    .withBuyerSecurityDeposit(offer.getBuyerSecurityDeposit().value)
                    .withPaymentAccountId(offer.getMakerPaymentAccountId())
                    .withPaymentMethodId(offer.getPaymentMethod().getId())
                    .withPaymentMethodShortName(offer.getPaymentMethod().getShortName())
                    .withBaseCurrencyCode(offer.getOfferPayload().getBaseCurrencyCode())
                    .withCounterCurrencyCode(offer.getOfferPayload().getCounterCurrencyCode())
                    .withDate(offer.getDate().getTime())
                    .build();
            CreateOfferReply reply = CreateOfferReply.newBuilder().setOffer(offerInfo.toProtoMessage()).build();
            responseObserver.onNext(reply);
            responseObserver.onCompleted();
}
```

Reply:
Imagine my embarrassment... but I'm still stuck.  The resultHandler passed into coreApi.createOffer has not been initialized when all is moved to the handler directly, and it cannot compile.

 BTW, this file was changed again in commit 82b7b79.

Reply:
Ah ok. I checked out the branch now to see more closely what the problem is.
We combined the offer creation and asycn offer placing in one method. This should be changed, and then the problem in the API code is also solved.

I will post below a patch, so its easier to see it in code... I just the did the minimum, need to be done on the other offer methods as well (we should also name it more clearly to make it easier understandable why there are 2 methods. I assume one is buyOffer and one is sellOffer (have not looked more in details yet)... the old domain code will require some refactorings as well... I hope I find in a few weeks time to join in those efforts...

Reply:
Index: core/src/main/java/bisq/core/api/CoreOffersService.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/src/main/java/bisq/core/api/CoreOffersService.java	(revision 939dba352b10d9f8049e7fea54a461ac93c777e2)
+++ core/src/main/java/bisq/core/api/CoreOffersService.java	(date 1601242413000)
@@ -77,7 +77,7 @@
         return offers;
     }
 
-    Offer createOffer(String currencyCode,
+    Offer getNewOffer(String currencyCode,
                       String directionAsString,
                       long priceAsLong,
                       boolean useMarketBasedPrice,
@@ -85,47 +85,16 @@
                       long amountAsLong,
                       long minAmountAsLong,
                       double buyerSecurityDeposit,
-                      String paymentAccountId,
-                      TransactionResultHandler resultHandler) {
+                      String paymentAccountId) {
         String offerId = createOfferService.getRandomOfferId();
         OfferPayload.Direction direction = OfferPayload.Direction.valueOf(directionAsString);
         Price price = Price.valueOf(currencyCode, priceAsLong);
         Coin amount = Coin.valueOf(amountAsLong);
         Coin minAmount = Coin.valueOf(minAmountAsLong);
         PaymentAccount paymentAccount = user.getPaymentAccount(paymentAccountId);
-        // We don't support atm funding from external wallet to keep it simple
-        boolean useSavingsWallet = true;
 
-        //noinspection ConstantConditions
-        return createAndPlaceOffer(offerId,
-                currencyCode,
-                direction,
-                price,
-                useMarketBasedPrice,
-                marketPriceMargin,
-                amount,
-                minAmount,
-                buyerSecurityDeposit,
-                paymentAccount,
-                useSavingsWallet,
-                resultHandler);
-    }
-
-    Offer createAndPlaceOffer(String offerId,
-                              String currencyCode,
-                              OfferPayload.Direction direction,
-                              Price price,
-                              boolean useMarketBasedPrice,
-                              double marketPriceMargin,
-                              Coin amount,
-                              Coin minAmount,
-                              double buyerSecurityDeposit,
-                              PaymentAccount paymentAccount,
-                              boolean useSavingsWallet,
-                              TransactionResultHandler resultHandler) {
         Coin useDefaultTxFee = Coin.ZERO;
-
-        Offer offer = createOfferService.createAndGetOffer(offerId,
+        return createOfferService.createAndGetOffer(offerId,
                 direction,
                 currencyCode,
                 amount,
@@ -136,18 +105,21 @@
                 marketPriceMargin,
                 buyerSecurityDeposit,
                 paymentAccount);
+    }
 
+    void placeOffer(Offer offer,
+                    double buyerSecurityDeposit,
+                    boolean useSavingsWallet,
+                    TransactionResultHandler resultHandler) {
         // TODO give user chance to examine offer before placing it (placeoffer)
         openOfferManager.placeOffer(offer,
                 buyerSecurityDeposit,
                 useSavingsWallet,
                 resultHandler,
                 log::error);
-
-        return offer;
     }
 
-    Offer createOffer(String offerId,
+    Offer getNewOffer(String offerId,
                       String currencyCode,
                       OfferPayload.Direction direction,
                       Price price,
Index: core/src/main/java/bisq/core/api/CoreApi.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/src/main/java/bisq/core/api/CoreApi.java	(revision 939dba352b10d9f8049e7fea54a461ac93c777e2)
+++ core/src/main/java/bisq/core/api/CoreApi.java	(date 1601242413000)
@@ -87,7 +87,7 @@
         return coreOffersService.getOffers(direction, fiatCurrencyCode);
     }
 
-    public Offer createOffer(String currencyCode,
+    public Offer getNewOffer(String currencyCode,
                              String directionAsString,
                              long priceAsLong,
                              boolean useMarketBasedPrice,
@@ -95,9 +95,8 @@
                              long amountAsLong,
                              long minAmountAsLong,
                              double buyerSecurityDeposit,
-                             String paymentAccountId,
-                             TransactionResultHandler resultHandler) {
-        return coreOffersService.createOffer(currencyCode,
+                             String paymentAccountId) {
+        return coreOffersService.getNewOffer(currencyCode,
                 directionAsString,
                 priceAsLong,
                 useMarketBasedPrice,
@@ -105,11 +104,20 @@
                 amountAsLong,
                 minAmountAsLong,
                 buyerSecurityDeposit,
-                paymentAccountId,
+                paymentAccountId);
+    }
+
+    public void placeOffer(Offer offer,
+                           double buyerSecurityDeposit,
+                           boolean useSavingsWallet,
+                           TransactionResultHandler resultHandler) {
+        coreOffersService.placeOffer(offer,
+                buyerSecurityDeposit,
+                useSavingsWallet,
                 resultHandler);
     }
 
-    public Offer createOffer(String offerId,
+    public Offer getNewOffer(String offerId,
                              String currencyCode,
                              OfferPayload.Direction direction,
                              Price price,
@@ -121,7 +129,7 @@
                              PaymentAccount paymentAccount,
                              boolean useSavingsWallet,
                              TransactionResultHandler resultHandler) {
-        return coreOffersService.createOffer(offerId,
+        return coreOffersService.getNewOffer(offerId,
                 currencyCode,
                 direction,
                 price,
Index: daemon/src/main/java/bisq/daemon/grpc/GrpcOffersService.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- daemon/src/main/java/bisq/daemon/grpc/GrpcOffersService.java	(revision 939dba352b10d9f8049e7fea54a461ac93c777e2)
+++ daemon/src/main/java/bisq/daemon/grpc/GrpcOffersService.java	(date 1601242365000)
@@ -20,7 +20,6 @@
 import bisq.core.api.CoreApi;
 import bisq.core.api.model.OfferInfo;
 import bisq.core.offer.Offer;
-import bisq.core.trade.handlers.TransactionResultHandler;
 
 import bisq.proto.grpc.CreateOfferReply;
 import bisq.proto.grpc.CreateOfferRequest;
@@ -35,7 +34,6 @@
 import javax.inject.Inject;
 
 import java.util.List;
-import java.util.concurrent.CountDownLatch;
 import java.util.stream.Collectors;
 
 import lombok.extern.slf4j.Slf4j;
@@ -90,12 +88,9 @@
     @Override
     public void createOffer(CreateOfferRequest req,
                             StreamObserver<CreateOfferReply> responseObserver) {
-        CountDownLatch latch = new CountDownLatch(1);
         try {
-            TransactionResultHandler resultHandler = transaction -> {
-                latch.countDown();
-            };
-            Offer offer = coreApi.createOffer(
+            // TODO give user chance to examine offer before placing it (placeoffer)
+            Offer offer = coreApi.getNewOffer(
                     req.getCurrencyCode(),
                     req.getDirection(),
                     req.getPrice(),
@@ -104,35 +99,37 @@
                     req.getAmount(),
                     req.getMinAmount(),
                     req.getBuyerSecurityDeposit(),
-                    req.getPaymentAccountId(),
-                    resultHandler);
-            try {
-                latch.await();
-            } catch (InterruptedException ignored) {
-                // empty
-            }
+                    req.getPaymentAccountId());
 
-            OfferInfo offerInfo = new OfferInfo.OfferInfoBuilder()
-                    .withId(offer.getId())
-                    .withDirection(offer.getDirection().name())
-                    .withPrice(offer.getPrice().getValue())
-                    .withUseMarketBasedPrice(offer.isUseMarketBasedPrice())
-                    .withMarketPriceMargin(offer.getMarketPriceMargin())
-                    .withAmount(offer.getAmount().value)
-                    .withMinAmount(offer.getMinAmount().value)
-                    .withVolume(offer.getVolume().getValue())
-                    .withMinVolume(offer.getMinVolume().getValue())
-                    .withBuyerSecurityDeposit(offer.getBuyerSecurityDeposit().value)
-                    .withPaymentAccountId(offer.getMakerPaymentAccountId())
-                    .withPaymentMethodId(offer.getPaymentMethod().getId())
-                    .withPaymentMethodShortName(offer.getPaymentMethod().getShortName())
-                    .withBaseCurrencyCode(offer.getOfferPayload().getBaseCurrencyCode())
-                    .withCounterCurrencyCode(offer.getOfferPayload().getCounterCurrencyCode())
-                    .withDate(offer.getDate().getTime())
-                    .build();
-            CreateOfferReply reply = CreateOfferReply.newBuilder().setOffer(offerInfo.toProtoMessage()).build();
-            responseObserver.onNext(reply);
-            responseObserver.onCompleted();
+            // We don't support atm funding from external wallet to keep it simple
+            boolean useSavingsWallet = true;
+
+            coreApi.placeOffer(offer,
+                    req.getBuyerSecurityDeposit(),
+                    useSavingsWallet,
+                    transaction -> {
+                        OfferInfo offerInfo = new OfferInfo.OfferInfoBuilder()
+                                .withId(offer.getId())
+                                .withDirection(offer.getDirection().name())
+                                .withPrice(offer.getPrice().getValue())
+                                .withUseMarketBasedPrice(offer.isUseMarketBasedPrice())
+                                .withMarketPriceMargin(offer.getMarketPriceMargin())
+                                .withAmount(offer.getAmount().value)
+                                .withMinAmount(offer.getMinAmount().value)
+                                .withVolume(offer.getVolume().getValue())
+                                .withMinVolume(offer.getMinVolume().getValue())
+                                .withBuyerSecurityDeposit(offer.getBuyerSecurityDeposit().value)
+                                .withPaymentAccountId(offer.getMakerPaymentAccountId())
+                                .withPaymentMethodId(offer.getPaymentMethod().getId())
+                                .withPaymentMethodShortName(offer.getPaymentMethod().getShortName())
+                                .withBaseCurrencyCode(offer.getOfferPayload().getBaseCurrencyCode())
+                                .withCounterCurrencyCode(offer.getOfferPayload().getCounterCurrencyCode())
+                                .withDate(offer.getDate().getTime())
+                                .build();
+                        CreateOfferReply reply = CreateOfferReply.newBuilder().setOffer(offerInfo.toProtoMessage()).build();
+                        responseObserver.onNext(reply);
+                        responseObserver.onCompleted();
+                    });
         } catch (IllegalStateException | IllegalArgumentException cause) {
             var ex = new StatusRuntimeException(Status.UNKNOWN.withDescription(cause.getMessage()));
             responseObserver.onError(ex);

Reply:
I did a refactoring to fix the tx result handler issue in commit 94996a5.  I'm working in a sub-branch, a child [PR 4559](https://github.com/bisq-network/bisq/pull/4559).